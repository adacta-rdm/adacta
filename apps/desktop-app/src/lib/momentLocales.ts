import moment from "moment";

// Bundle de
import "moment/dist/locale/de.js";
moment.locale("de");
