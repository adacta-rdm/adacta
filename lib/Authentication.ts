import { pbkdf2 as _pbkdf2, randomBytes } from "crypto";
import { promisify } from "util";

const pbkdf2 = promisify(_pbkdf2);

const hashIterations = 100000;

async function hashPassword(password: string, salt: string) {
	const buffer = await pbkdf2(password, salt, hashIterations, 64, "sha512");
	return buffer.toString("base64");
}

export async function verifyPassword(password: string, hash: string, salt: string) {
	const newHash = await hashPassword(password, salt);
	return newHash === hash;
}

export async function createPasswordHash(password: string) {
	const salt = randomBytes(128).toString("base64");
	const passwordHash = await hashPassword(password, salt);

	return { salt, passwordHash };
}
