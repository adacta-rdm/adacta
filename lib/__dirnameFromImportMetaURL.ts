/**
 * Pass in `import.meta.url` as the argument and it will return the directory name of the current, similar to
 * `__dirname` in commonjs.
 *
 * Avoids the use of Node APIs so that it also works in the browser.
 */
export function __dirnameFromImportMetaURL(importMetaURL: string): string {
	const { pathname } = new URL(importMetaURL);
	const lastSlash = pathname.lastIndexOf("/");

	return pathname.slice(0, lastSlash);
}
