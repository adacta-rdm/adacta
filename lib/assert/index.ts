/**
 * @file Automatically generated by barrelsby.
 */

export * from "./assertDefined";
export * from "./assertInstanceof";
export * from "./assertUnreachable";
export * from "./isNonNullish";
