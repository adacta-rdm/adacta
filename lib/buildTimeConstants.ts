/**
 * The constants defined in this file are embedded into the bundles at build time using the __ADACTA global.
 * During development, the __ADACTA global is not defined, so the constants are defined dynamically.
 *
 * This approach allows us to also bundle text files into the build, while maintaining compatibility with tsx which is
 * used in development (while tsx uses esbuild under the hood, it does not support esbuild loaders).
 */
import { readFileSync } from "node:fs";

// NOTE: This import should not be shortened to "~/dev/sh" because it is used in the build process.
import { shSync } from "../dev/sh";
import { version } from "../package.json";

globalThis.__ADACTA = globalThis.__ADACTA ?? {};

export const COMMIT_SHA =
	__ADACTA.COMMIT_SHA ?? shSync("git rev-parse --short HEAD").trim().slice(0, 8);
export const BUILD_DATE = __ADACTA.BUILD_DATE ?? new Date().toISOString().slice(0, 10);

/**
 * The version specified in the package.json file.
 */
export const PACKAGE_VERSION = __ADACTA.PACKAGE_VERSION ?? version;

export const TYPE_DEFS =
	__ADACTA.TYPE_DEFS ??
	readFileSync(`${__dirname}/../apps/repo-server/src/graphql/generated/schema.graphql`, "utf-8");

export const INIT_REPO_SQL =
	__ADACTA.INIT_REPO_SQL ?? readFileSync(`${__dirname}/../drizzle/init-repo.sql`, "utf-8");

export const INIT_MIGRATIONS_SQL =
	__ADACTA.INIT_MIGRATIONS_SQL ??
	readFileSync(`${__dirname}/../drizzle/init-migrations.sql`, "utf-8");

declare global {
	// The value of the Record type is typed as undefined, so the actually exported value gets the type of the
	// default value (i.e. the part after the ?? operator).
	// This needs to be declared using var...
	// eslint-disable-next-line no-var
	var __ADACTA: Record<string, undefined>;
}
