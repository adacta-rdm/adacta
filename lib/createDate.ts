import type { Opaque } from "type-fest";

const ISO_REGEX =
	/(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))/;

export type IDatetime = Opaque<string, Date>;

export function assertIsoDate(isoString: string): asserts isoString is IDatetime {
	if (!ISO_REGEX.test(isoString)) {
		throw new Error(`'${isoString}' is not an ISO Date string`);
	}
}

export function createMaybeDate(maybeIsoString: string | null | undefined) {
	if (maybeIsoString === null || maybeIsoString === undefined) {
		return undefined;
	}
	return createDate(maybeIsoString);
}

export function createDate(isoString: string) {
	assertIsoDate(isoString);
	return new Date(isoString);
}

export function createMaybeIDatetime(
	date: Date | IDatetime | null | undefined
): IDatetime | undefined {
	if (date === null || date === undefined) {
		return undefined;
	}
	return createIDatetime(date);
}

export function createIDatetime(date: Date | IDatetime): IDatetime {
	if (date instanceof Date) return date.toISOString() as IDatetime;
	return date;
}
