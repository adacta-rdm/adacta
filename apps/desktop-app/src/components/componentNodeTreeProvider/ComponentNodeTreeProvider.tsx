/* eslint-disable relay/unused-fields */
import assert from "assert";

import { groupBy } from "lodash";
import type { ReactNode } from "react";
import React, { createContext } from "react";
import { graphql, useFragment } from "react-relay";
import type { ArrayElement } from "type-fest/source/internal";

import { propertyNameToReadableString } from "../../utils/PropertyNameToReadableString";

import type {
	ComponentNodeTreeProviderFragment$data,
	ComponentNodeTreeProviderFragment$key,
} from "@/relay/ComponentNodeTreeProviderFragment.graphql";
import type { DeviceListHierarchicalGraphQLFragment$data } from "@/relay/DeviceListHierarchicalGraphQLFragment.graphql";
import { splitPropertyNameIntoVirtualGroups } from "~/lib/utils/splitPropertyNameIntoVirtualGroups";

const ComponentNodeTreeProviderFragmentGraphQl = graphql`
	fragment ComponentNodeTreeProviderFragment on Device
	@argumentDefinitions(timeFrame: { type: "TimeFrameInput" }, time: { type: "DateTime" }) {
		components(time: $time, timeFrame: $timeFrame, includeOverlaps: true) {
			component {
				__typename
				... on Device {
					id
					name
					displayName
					usagesAsProperty(time: $time, timeFrame: $timeFrame, includeOverlaps: true) {
						id
						device {
							id
						}
						timestamp
						timestampEnd
						name
						value {
							# eslint-disable-next-line relay/must-colocate-fragment-spreads
							...DeviceLink
						}
					}
					definition {
						acceptsUnit
					}
				}
				... on Sample {
					id
					name
					usagesAsProperty(time: $time, timeFrame: $timeFrame, includeOverlaps: true) {
						id
						device {
							id
						}
						timestamp
						timestampEnd
						name
					}
				}
			}
			# eslint-disable-next-line relay/unused-fields
			pathFromTopLevelDevice
		}
	}
`;

interface IProps {
	device: ComponentNodeTreeProviderFragment$key;
	children: ReactNode;
}

interface IContext<T extends ComponentNodeTreeProviderFragmentDataUnion> {
	tree: IComponentTreeNode<T>[];
	components: TComponentOutput<ComponentNodeTreeProviderFragment$data>[];
}

type ComponentNodeTreeProviderFragmentDataUnion =
	| ComponentNodeTreeProviderFragment$data
	| DeviceListHierarchicalGraphQLFragment$data;

/**
 * Generic tree structure that different components can wrap however they want.
 * E.g. the `DeviceOverview` component creates a `EuiTree` while the `SetupDescription` component creates a flat list
 * of tags.
 */
export interface IComponentTreeNode<
	TDatasource extends ComponentNodeTreeProviderFragmentDataUnion
> {
	name: string;
	tag: string; // e.g. 3.1.2
	component: TComponentOutput<TDatasource>["component"];
	children: IComponentTreeNode<TDatasource>[];
}

type TComponent<TDatasource extends ComponentNodeTreeProviderFragmentDataUnion> = ArrayElement<
	TDatasource["components"]
>;

type TComponentOutput<TDatasource extends ComponentNodeTreeProviderFragmentDataUnion> =
	| TComponent<TDatasource>
	| {
			pathFromTopLevelDevice: string[];
			readonly component: { readonly __typename: "virtualGroup"; name?: "Test" };
	  };

function createVirtualGroups<TDatasource extends ComponentNodeTreeProviderFragmentDataUnion>(
	components: TComponentOutput<TDatasource>[]
): TComponentOutput<TDatasource>[] {
	type TComponentBound = TComponentOutput<TDatasource>;

	const virtualGroupsCreated: string[] = [];

	return [...components].flatMap<TComponentBound>((component): TComponentBound[] => {
		const oldPath = component.pathFromTopLevelDevice;
		const newPath = splitPropertyNameIntoVirtualGroups(oldPath);

		const c: TComponentBound = {
			...component,
			pathFromTopLevelDevice: newPath,
		};

		// If the path changes when splitPropertyNameIntoVirtualGroups is applied then this should
		// be a virtual group
		if (newPath.length !== oldPath.length) {
			const newParentPath = newPath.slice(0, -1);

			const identifier = JSON.stringify(newParentPath);
			if (!virtualGroupsCreated.includes(identifier)) {
				const virtualGroup: TComponentBound = {
					pathFromTopLevelDevice: newParentPath,
					component: { __typename: "virtualGroup" },
				};

				virtualGroupsCreated.push(identifier);

				// Return the virtual group and the component
				return [c, virtualGroup];
			}
		}

		return [c];
	});
}

export function createComponentNodeTree<
	TDatasource extends ComponentNodeTreeProviderFragmentDataUnion
>(c: Readonly<TComponent<TDatasource>[]>, depthString = ""): IComponentTreeNode<TDatasource>[] {
	// Group by first entry of property path. each group with more than one entry becomes a new "folder" in the
	// tree. The resulting object looks like this:
	// {
	//    mfc1: [ {	device: {...}, pathFromTopLevelDevice: ["mfc1"] } ]
	//    ...
	//    cem1: [
	//	    { device: {...}, pathFromTopLevelDevice: ["cem1"] }
	//	    { device: {...}, pathFromTopLevelDevice: ["cem1", "lfc"] }
	//	    { device: {...}, pathFromTopLevelDevice: ["cem1", "mfc"] }
	//    ]
	//    ...
	// }

	const components: TComponentOutput<TDatasource>[] = createVirtualGroups([...c]);
	const grouped = groupBy(components, (d) => d.pathFromTopLevelDevice[0]);

	return (
		Object.values(grouped)
			// Note: Double sort is important to avoid results like this: "Mfc 1", "Mfc 10", "Mfc 2", ...
			// Sorts in the following way:
			// 1. All properties that do not end with a number in alphabetical order.
			// 2. All properties that end with a number:
			//  2.1 Sorted by the whole property name first
			//  2.2 Within the same property name they are sorted by the number at the end
			.sort((a, b) => {
				return propertyNameToReadableString(a[0].pathFromTopLevelDevice[0]).localeCompare(
					propertyNameToReadableString(b[0].pathFromTopLevelDevice[0])
				);
			})
			.sort((a, b) => {
				const endNumberA = Number(
					propertyNameToReadableString(a[0].pathFromTopLevelDevice[0]).split(" ").slice(-1)[0]
				);
				const endNumberB = Number(
					propertyNameToReadableString(b[0].pathFromTopLevelDevice[0]).split(" ").slice(-1)[0]
				);
				if (isNaN(endNumberA) && !isNaN(endNumberB)) return -1;
				if (isNaN(endNumberB) && !isNaN(endNumberA)) return 1;
				if (!isNaN(endNumberA) && !isNaN(endNumberB))
					return Number(endNumberA) - Number(endNumberB);
				return propertyNameToReadableString(a[0].pathFromTopLevelDevice[0]).localeCompare(
					propertyNameToReadableString(b[0].pathFromTopLevelDevice[0])
				);
			})
			.flatMap((subComponentsWithPath: TComponentOutput<TDatasource>[], i) => {
				const topLevelSubComponentsWithPath: TComponentOutput<TDatasource>[] =
					subComponentsWithPath.filter((sc) => sc.pathFromTopLevelDevice.length === 1);

				const results: IComponentTreeNode<TDatasource>[] = [];

				for (const topLevelSubComponentWithPath of topLevelSubComponentsWithPath) {
					const component = topLevelSubComponentWithPath.component;

					const tag = `${depthString}${String(i + 1)}`;
					assert(component.__typename !== "%other");

					if (topLevelSubComponentWithPath.pathFromTopLevelDevice.length === 0) {
						continue;
					}

					// TODO: Remove propertyNameToReadableString
					const name = propertyNameToReadableString(
						topLevelSubComponentWithPath.pathFromTopLevelDevice[0]
					);

					const children =
						subComponentsWithPath.length === 1
							? []
							: createComponentNodeTree(
									subComponentsWithPath
										.filter((sd) => sd.pathFromTopLevelDevice.length > 1)
										.map((sd) => ({
											...sd,
											pathFromTopLevelDevice: sd.pathFromTopLevelDevice.slice(1),
										})) as TComponent<TDatasource>[],
									`${depthString + String(i + 1)}.`
							  );

					// Casting here is necessary because the type of the children property is
					// somehow not correctly inferred (and is never)
					// I think it is related to the recursive nature of the type or something
					// regarding the difference between `(A|B)[]` and `A[] | B[]`
					const newElement: IComponentTreeNode<TDatasource> = {
						name,
						tag,
						component: component,
						children: children,
					} as IComponentTreeNode<TDatasource>;

					results.push(newElement);
				}
				return results;
			})
	);
}

const TreeContext = createContext<IContext<ComponentNodeTreeProviderFragment$data> | null>(null);

export function ComponentNodeTreeProvider(props: IProps) {
	const data = useFragment(ComponentNodeTreeProviderFragmentGraphQl, props.device);
	const tree = createComponentNodeTree<ComponentNodeTreeProviderFragment$data>(data.components);

	const components = createVirtualGroups<ComponentNodeTreeProviderFragment$data>([
		...data.components,
	]);

	return (
		<TreeContext.Provider value={{ tree, components: components }} {...props}>
			{props.children}
		</TreeContext.Provider>
	);
}

export function useTree() {
	const context = React.useContext(TreeContext);
	if (!context) {
		throw new Error(`useTree must be used within a ComponentNodeTreeProvider`);
	}
	return context;
}
