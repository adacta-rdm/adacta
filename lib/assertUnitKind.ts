import { AssertionError } from "assert";

import { Qty } from "@omegadot/einheiten";
import type { UnitKind } from "@omegadot/einheiten/dist/types/quantities/kind";
import { difference } from "lodash";

export function assertUnitKind(kind: string): asserts kind is UnitKind {
	if (!Qty.getKinds().includes(kind as UnitKind)) {
		throw new AssertionError({ message: `Unit kind ${kind} is not a valid unit kind` });
	}
}

export function assertUnitKinds(kind: string[] | Readonly<string[]>): asserts kind is UnitKind[] {
	const inputUnique = [...new Set(kind)];
	const validKinds = Qty.getKinds();

	const invalidUnits = difference(inputUnique, validKinds);

	if (invalidUnits.length > 0) {
		throw new AssertionError({
			message: `The following unit kinds are not valid unit kinds: ${invalidUnits.join(",")}`,
		});
	}
}
