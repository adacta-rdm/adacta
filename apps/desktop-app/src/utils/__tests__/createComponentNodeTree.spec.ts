import type { FragmentRefs } from "relay-runtime";
import { expect, test } from "vitest";

import { createComponentNodeTree } from "../../components/componentNodeTreeProvider/ComponentNodeTreeProvider";

import type { ComponentNodeTreeProviderFragment$data } from "@/relay/ComponentNodeTreeProviderFragment.graphql";
import { createIDatetime } from "~/lib/createDate";

test("createComponentNodeTree", () => {
	// The value property is not used by the function under test, so we can omit it.
	const fakeRef = {
		value: null as unknown as {
			readonly " $fragmentSpreads": FragmentRefs<"DeviceLink">;
		},
	};

	const components: ComponentNodeTreeProviderFragment$data["components"] = [
		{
			component: {
				__typename: "Sample",
				id: "s",
				name: "PL-Pd-Pt-CZ",
				usagesAsProperty: [
					{
						name: "sample",
						timestamp: createIDatetime(new Date(0)),
						timestampEnd: createIDatetime(new Date(0)),
						id: "fakePropertyId",
						device: { id: "fakeDeviceId" },
						...fakeRef,
					},
				],
			},
			pathFromTopLevelDevice: ["furnace", "tube", "sample"],
		},
		{
			component: {
				__typename: "Device",
				id: "t",
				name: "Quartz tube",
				displayName: "Quartz tube",
				usagesAsProperty: [
					{
						name: "tube",
						timestamp: createIDatetime(new Date(0)),
						timestampEnd: createIDatetime(new Date(0)),
						id: "fakePropertyId",
						device: { id: "fakeDeviceId" },
						...fakeRef,
					},
				],
				definition: {
					acceptsUnit: [],
				},
			},
			pathFromTopLevelDevice: ["furnace", "tube"],
		},
		{
			component: {
				__typename: "Device",
				id: "f",
				name: "My furnace",
				displayName: "My furnace",
				usagesAsProperty: [
					{
						name: "furnace",
						timestamp: createIDatetime(new Date(0)),
						timestampEnd: createIDatetime(new Date(0)),
						id: "fakePropertyId",
						device: { id: "fakeDeviceId" },
						...fakeRef,
					},
				],
				definition: {
					acceptsUnit: [],
				},
			},
			pathFromTopLevelDevice: ["furnace"],
		},
		{
			component: {
				__typename: "Device",
				id: "m",
				name: "mfc x",
				displayName: "mfc x",
				usagesAsProperty: [
					{
						name: "mfc1",
						timestamp: createIDatetime(new Date(0)),
						timestampEnd: createIDatetime(new Date(0)),
						id: "fakePropertyId",
						device: { id: "fakeDeviceId" },
						...fakeRef,
					},
				],
				definition: {
					acceptsUnit: [],
				},
			},
			pathFromTopLevelDevice: ["mfc1"],
		},
	];
	expect(createComponentNodeTree(components)).toMatchSnapshot();
});
