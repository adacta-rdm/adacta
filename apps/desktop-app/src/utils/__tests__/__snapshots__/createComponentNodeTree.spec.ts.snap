// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createComponentNodeTree 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [],
            "component": {
              "__typename": "Sample",
              "id": "s",
              "name": "PL-Pd-Pt-CZ",
              "usagesAsProperty": [
                {
                  "device": {
                    "id": "fakeDeviceId",
                  },
                  "id": "fakePropertyId",
                  "name": "sample",
                  "timestamp": "1970-01-01T00:00:00.000Z",
                  "timestampEnd": "1970-01-01T00:00:00.000Z",
                  "value": null,
                },
              ],
            },
            "name": "Sample",
            "tag": "1.1.1",
          },
        ],
        "component": {
          "__typename": "Device",
          "definition": {
            "acceptsUnit": [],
          },
          "displayName": "Quartz tube",
          "id": "t",
          "name": "Quartz tube",
          "usagesAsProperty": [
            {
              "device": {
                "id": "fakeDeviceId",
              },
              "id": "fakePropertyId",
              "name": "tube",
              "timestamp": "1970-01-01T00:00:00.000Z",
              "timestampEnd": "1970-01-01T00:00:00.000Z",
              "value": null,
            },
          ],
        },
        "name": "Tube",
        "tag": "1.1",
      },
    ],
    "component": {
      "__typename": "Device",
      "definition": {
        "acceptsUnit": [],
      },
      "displayName": "My furnace",
      "id": "f",
      "name": "My furnace",
      "usagesAsProperty": [
        {
          "device": {
            "id": "fakeDeviceId",
          },
          "id": "fakePropertyId",
          "name": "furnace",
          "timestamp": "1970-01-01T00:00:00.000Z",
          "timestampEnd": "1970-01-01T00:00:00.000Z",
          "value": null,
        },
      ],
    },
    "name": "Furnace",
    "tag": "1",
  },
  {
    "children": [],
    "component": {
      "__typename": "Device",
      "definition": {
        "acceptsUnit": [],
      },
      "displayName": "mfc x",
      "id": "m",
      "name": "mfc x",
      "usagesAsProperty": [
        {
          "device": {
            "id": "fakeDeviceId",
          },
          "id": "fakePropertyId",
          "name": "mfc1",
          "timestamp": "1970-01-01T00:00:00.000Z",
          "timestampEnd": "1970-01-01T00:00:00.000Z",
          "value": null,
        },
      ],
    },
    "name": "Mfc 1",
    "tag": "2",
  },
]
`;
